doc/api/n-api.md

```diff
@@ -3155,6 +3156,81 @@ support it:
 * If the function is not available, provide an alternate implementation
   that does not use the function.
 
+## Promises
+
+Promises facilitate the construction of chains of asynchronous actions such
+that each action hinges on the outcome of the previous action.
+
+N-API implements promises as a pair of two objects - a "deferred" and a
+promise. A "deferred" is a JavaScript value created along with the promise by
+`napi_create_promise()`. The promise can be passed into JavaScript while the
+deferred must be retained on the native side, because the promise that has been
+derived from it can only be resolved or rejected by passing the deferred into
+`napi_conclude_deferred()`.
+
+### napi_create_promise
+<!-- YAML
+added: REPLACEME
+-->
+```C
+NAPI_EXTERN napi_status napi_create_promise(napi_env env,
+                                            napi_value* deferred,
+                                            napi_value* promise);
+```
+
+- `[in] env`: The environment that the API is invoked under.
+- `[out] deferred`: The deferred object.
+- `[out] promise`: The JavaScript promise associated with the deferred object.
+
+Returns `napi_ok` if the API succeeded.
+
+This API creates a deferred object and a JavaScript promise. The two objects
+are linked such that the deferred can later be passed to
+`napi_conclude_deferred()` to resolve the promise. Although the deferred object
+is a JavaScript value it should not be passed into JavaScript because it lacks
+the functionality of a JavaScript promise. However, the JavaScript promise
+returned in the third parameter can be passed into JavaScript.
+
+### napi_conclude_deferred
+<!-- YAML
+added: REPLACEME
+-->
+```C
+NAPI_EXTERN napi_status napi_conclude_deferred(napi_env env,
+                                               napi_value deferred,
+                                               napi_value conclusion,
+                                               bool is_resolution);
+```
+
+- `[in] env`: The environment that the API is invoked under.
+- `[in] deferred`: The deferred object whose associated promise to conclude.
+- `[in] conclusion`: The value with which to conclude the promise.
+- `[in] is_rejection`: Flag indicating whether `conclusion` is a resolution
```

Comment:
Wouldn't it be better to have `napi_promise_fulfill(env, deferred, value)` and `napi_promise_reject(env, deferred, value)`?

When I think about it - why do we even need a deferred object at all?